### PR TITLE
[RFR] Feature/google cloud client adapter

### PR DESCRIPTION
**Pull request by nicolasmure**
Supersede #427 .

## TODO
- [x] update doc

### CODE REVIEW

**Comment by cdaguerre** [comment]
Any reason why this wasn't merged?

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    private $bucket;
+    private $options      = array();
+    private $metadata     = array();
+
+    /**
+     * @param StorageClient    $service    Authenticated storage client class
+     * @param string           $bucketName Name of the bucket
+     * @param array            $options    Options are: "directory" and "acl" (see https://cloud.google.com/storage/docs/access-control/lists)
+     */
+    public function __construct(StorageClient $storageClient, $bucketName, $options = array())
```
Comment (NiR-) [inline src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: Can you typehint `$bucketName` please? 1.x branch requires php >= 7.1

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    {
+        return $this->exists($this->computePath(rtrim($key, '/')).'/');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listKeys($prefix = null)
+    {
+        // @FIXME : the list should not return the root directory in the keys
```
Comment (NiR-) [inline src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: Is this fixed already?

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        $keys = array();
+
+        foreach ($this->bucket->objects(array('prefix' => $this->computePath($prefix))) as $e) {
```
Comment (NiR-) [inline src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: Lots of parenthesis here, could you put `objects()` result in a variable please? And maybe switch to newer array notation

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+            throw StorageFailure::unexpectedFailure('delete', ['key' => $key], $e);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function rename($sourceKey, $targetKey)
+    {
+        $pathedSourceKey = $this->computePath($sourceKey);
```
Comment (NiR-) [inline src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: `pathed` is strange IMHO, maybe use `resolved` instead? :thinking:

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMetadata($key)
+    {
+        try {
+            $infos = $this->bucket->object($this->computePath($key))->info();
+
+            return isset($infos['metadata'])
```
Comment (NiR-) [inline src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: Maybe use the coalesce operator here :thinking: (and maybe use a cs-fixer to fix everything :stuck_out_tongue:)

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setMetadata($key, $metadata)
+    {
+        try {
+            $this->bucket->object($this->computePath($key))->update(array('metadata' => $metadata));
```
Comment (NiR-) [inline src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: Could you write this on multiple line please?

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+    /**
+     * @param StorageClient    $service    Authenticated storage client class
+     * @param string           $bucketName Name of the bucket
+     * @param array            $options    Options are: "directory" and "acl" (see https://cloud.google.com/storage/docs/access-control/lists)
+     */
+    public function __construct(StorageClient $storageClient, $bucketName, $options = array())
+    {
+        $this->storageClient = $storageClient;
+        $this->initBucket($bucketName);
```
Comment (NiR-) [inline src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: I'm not sure about this one. I think it's done by other adapters too but it's totally questionable: not only a bucket has to be created but it also has to be properly configured (lifecycle & retention policies, IAM bindings, etc...) and this is way beyond Gaufrette feature set IMHO.

EDIT: Or I misread the code and this only initialize `bucket` property? :thinking:

File: tests/Gaufrette/Functional/Adapter/GoogleCloudClientStorageTest.php [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+{
+    private $string    = 'Yeah mate. No worries, I uploaded just fine. Meow!';
+    private $directory = 'tests';
+    private $bucketName;
+    private $sdkOptions;
+    private $bucketOptions;
+
+    public function setUp()
+    {
+        $gccsProjectId   = getenv('GCCS_PROJECT_ID');
```
Comment (NiR-) [inline tests/Gaufrette/Functional/Adapter/GoogleCloudClientStorageTest.php]: It's strange to call these variables GCCS_* and it's actually strange to call this adapter `GoogleCloudClientStorage`. I mean, there's no need for that `Client`.

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php (hunk 7 above)
Comment (nicolasmure) [reply src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: Yes it initializes the `bucket` property with a lazy objects from the SDK (no requests made to the backend), cf https://sourcegraph.com/github.com/googleapis/google-cloud-php/-/blob/Storage/src/StorageClient.php#L125

File: src/Gaufrette/Adapter/GoogleCloudClientStorage.php (hunk 2 above)
Comment (nicolasmure) [reply src/Gaufrette/Adapter/GoogleCloudClientStorage.php]: Yes, I added a test for it :)

**Pull request merged**